src/main/java/com/indeed/jiraactions/api/ApiCaller.java

```diff
@@ -3,173 +3,94 @@
+
+            if (!response.isSuccessful()) {
+                final StringBuilder sb = new StringBuilder();
+                Headers requestHeaders = request.headers();
+                Headers responseHeaders = response.headers();
+
+                sb.append("{");
+                sb.append("\"Request\": {");
+                sb.append("\"URL\": \"").append(url).append("\",");
 
-            sb.append("\"URL\": \"").append(url).append("\",");
-            if (headers != null) {
                 sb.append("\"Headers\": {");
-                for (final Map.Entry<String, List<String>> header : headers.entrySet()) {
-                    final String key = header.getKey();
+                for (int i = 0, size = requestHeaders.size(); i < size; i++) {
+                    final String key = requestHeaders.name(i);
                     final String value;
                     if ("Authorization".equals(key)) {
                         value = "<Omitted>";
                     } else {
-                        value = String.join(",", header.getValue());
+                        value = String.join(",", requestHeaders.value(i));
                     }
                     sb.append("\"").append(key).append("\": \"").append(value).append("\",");
                 }
-            }
-            sb.append("}");
-
-            sb.append("}");
+                sb.append("}");
+                sb.append("}");
 
-            if (urlConnection != null) {
                 sb.append(", \"Response\": {");
-                for (final Map.Entry<String, List<String>> header : urlConnection.getHeaderFields().entrySet()) {
-                    final String key = header.getKey();
+                for (int i = 0, size = responseHeaders.size(); i < size; i++) {
+                    final String key = responseHeaders.name(i);
                     final String value;
                     if ("Set-Cookie".equals(key)) {
                         value = "<Omitted>";
                     } else {
-                        value = String.join(",", header.getValue());
+                        value = String.join(",", responseHeaders.value(i));
                     }
                     sb.append("\"").append(key).append("\": \"").append(value).append("\",");
                 }
-                sb.append("\"Code\": ").append(urlConnection.getResponseCode()).append(",");
-                sb.append("\"Message\": \"").append(urlConnection.getResponseMessage()).append("\",");
+                sb.append("\"Code\": ").append(response.code()).append(",");
+                sb.append("\"Message\": \"").append(response.message()).append("\",");
+                sb.append("\"Error Body\": \"").append(responseBody.string()).append("\"");
+                sb.append("}");
+                sb.append("}");
+                log.debug("Encountered connection error: " + sb);
+                throw new IOException(String.valueOf(response));
```

Comment:
Generally you should either log a message or throw, but not both. If you do both, you end up with double logging (because whomever catches this exception will presumably also log an error too).

(Yes, I see the existing code logs and throws as well; it's still worth pointing out.)